File: R/streamgraph.R
#streamgraph R API. The graph returned by stream_graph() is an external pointer;
#pass it as the first argument to every function. Node IDs are 0-indexed.

#' streamgraph: streaming (dynamic) graph analytics
#'
#' Maintains connected components, triangle counts, degrees and basic centrality
#' incrementally as edges arrive. The graph object is an external pointer; node
#' IDs are non-negative integers, 0-indexed.
#'
#' @useDynLib streamgraph, .registration = TRUE
#' @importFrom Rcpp evalCpp
#' @keywords internal
"_PACKAGE"

#' Create a StreamGraph
#'
#' @param n_nodes Integer; 0 to auto-grow (default), or N to pre-allocate N
#'   node slots when an upper bound is known.
#' @param directed Logical; whether the graph is directed. Default FALSE.
#' @return An external pointer to the underlying C++ StreamGraph.
#' @examples
#' G <- stream_graph()
#' add_edge(G, 0L, 1L)
#' n_edges(G)
#' @export
stream_graph <- function(n_nodes = 0L, directed = FALSE) {
    .sg_create(as.integer(n_nodes), as.logical(directed))
}

#' Add a single edge
#'
#' @param G A StreamGraph (from [stream_graph()]).
#' @param u,v Node IDs (0-indexed). Storage auto-expands to fit.
#' @return Logical: TRUE if the edge was new, FALSE for a self-loop or duplicate.
#' @examples
#' G <- stream_graph()
#' add_edge(G, 0L, 1L)
#' add_edge(G, 0L, 1L)  # FALSE: duplicate
#' @export
add_edge <- function(G, u, v) {
    .sg_add_edge(G, as.integer(u), as.integer(v))
}

#' Add many edges in a batch
#'
#' Duplicates within the batch and self-loops are removed in a parallel pre-pass;
#' unique edges are then applied serially.
#'
#' @param G A StreamGraph.
#' @param us,vs Integer vectors of equal length holding 0-indexed endpoints.
#' @param n_threads Integer; OpenMP threads for the dedup sort (0 = default).
#' @return Numeric: the number of new edges added.
#' @examples
#' G <- stream_graph()
#' add_edges(G, c(0L, 1L, 2L), c(1L, 2L, 0L))
#' @export
add_edges <- function(G, us, vs, n_threads = 0L) {
    .sg_add_edges(G, as.integer(us), as.integer(vs), as.integer(n_threads))
}

#' Whether two nodes are in the same component
#'
#' @param G A StreamGraph.
#' @param u,v Node IDs (0-indexed).
#' @return Logical.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' same_component(G, 0L, 1L)
#' @export
same_component <- function(G, u, v) {
    .sg_same_component(G, as.integer(u), as.integer(v))
}

#' Component root ID of a node
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Integer root ID.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' component_id(G, 1L)
#' @export
component_id <- function(G, u) {
    .sg_component_id(G, as.integer(u))
}

#' Number of connected components (touched nodes only)
#'
#' @param G A StreamGraph.
#' @return Integer.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L); add_edge(G, 2L, 3L)
#' n_components(G)
#' @export
n_components <- function(G) {
    .sg_n_components(G)
}

#' Number of nodes in a node's component
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Integer.
#' @examples
#' G <- stream_graph(); for (i in 0:4) add_edge(G, i, i + 1L)
#' component_size(G, 0L)
#' @export
component_size <- function(G, u) {
    .sg_component_size(G, as.integer(u))
}

#' All touched nodes in a node's component
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Integer vector of node IDs (0-indexed).
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L); add_edge(G, 1L, 2L)
#' component_nodes(G, 0L)
#' @export
component_nodes <- function(G, u) {
    .sg_component_nodes(G, as.integer(u))
}

#' Component root ID for every touched node
#'
#' @param G A StreamGraph.
#' @return Integer vector, one entry per touched node in ascending ID order.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' component_ids(G)
#' @export
component_ids <- function(G) {
    .sg_component_ids(G)
}

#' Degree of a node
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Integer.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L); add_edge(G, 0L, 2L)
#' degree(G, 0L)
#' @export
degree <- function(G, u) {
    .sg_degree(G, as.integer(u))
}

#' Degree histogram
#'
#' @param G A StreamGraph.
#' @return Named numeric vector; entry "d" is the number of nodes of degree d.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L); add_edge(G, 0L, 2L)
#' degree_histogram(G)
#' @export
degree_histogram <- function(G) {
    .sg_degree_histogram(G)
}

#' Sorted neighbours of a node
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Sorted integer vector of neighbour IDs (0-indexed).
#' @examples
#' G <- stream_graph(); add_edge(G, 5L, 2L); add_edge(G, 5L, 1L)
#' neighbours(G, 5L)
#' @export
neighbours <- function(G, u) {
    .sg_neighbours(G, as.integer(u))
}

#' Whether an edge exists
#'
#' @param G A StreamGraph.
#' @param u,v Node IDs (0-indexed).
#' @return Logical.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' has_edge(G, 0L, 1L)
#' @export
has_edge <- function(G, u, v) {
    .sg_has_edge(G, as.integer(u), as.integer(v))
}

#' Number of touched (referenced) nodes
#'
#' @param G A StreamGraph.
#' @return Integer.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' n_nodes(G)
#' @export
n_nodes <- function(G) {
    .sg_n_nodes(G)
}

#' Number of edges
#'
#' @param G A StreamGraph.
#' @return Numeric.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' n_edges(G)
#' @export
n_edges <- function(G) {
    .sg_n_edges(G)
}

#' Global triangle count
#'
#' @param G A StreamGraph.
#' @return Numeric; each triangle counted once.
#' @examples
#' G <- stream_graph()
#' for (e in list(c(0,1), c(1,2), c(0,2))) add_edge(G, e[1], e[2])
#' triangle_count(G)
#' @export
triangle_count <- function(G) {
    .sg_triangle_count(G)
}

#' Triangles incident to a node
#'
#' @param G A StreamGraph.
#' @param u Node ID (0-indexed).
#' @return Integer.
#' @examples
#' G <- stream_graph()
#' for (e in list(c(0,1), c(1,2), c(0,2))) add_edge(G, e[1], e[2])
#' local_triangles(G, 0L)
#' @export
local_triangles <- function(G, u) {
    .sg_local_triangles(G, as.integer(u))
}

#' Per-node local triangle counts
#'
#' @param G A StreamGraph.
#' @return Integer vector indexed by node ID over the allocated range.
#' @examples
#' G <- stream_graph()
#' for (e in list(c(0,1), c(1,2), c(0,2))) add_edge(G, e[1], e[2])
#' all_local_triangles(G)
#' @export
all_local_triangles <- function(G) {
    .sg_all_local_triangles(G)
}

#' Edge density
#'
#' @param G A StreamGraph.
#' @return Numeric: 2m/(n(n-1)) undirected, m/(n(n-1)) directed.
#' @examples
#' G <- stream_graph(); for (i in 0:3) add_edge(G, i, i + 1L)
#' density(G)
#' @export
density <- function(G) {
    .sg_density(G)
}

#' Average degree
#'
#' @param G A StreamGraph.
#' @return Numeric: 2m/n undirected, m/n directed.
#' @examples
#' G <- stream_graph(); for (i in 0:3) add_edge(G, i, i + 1L)
#' avg_degree(G)
#' @export
avg_degree <- function(G) {
    .sg_avg_degree(G)
}

#' Maximum degree
#'
#' @param G A StreamGraph.
#' @return Integer.
#' @examples
#' G <- stream_graph(); for (i in 0:3) add_edge(G, i, i + 1L)
#' max_degree(G)
#' @export
max_degree <- function(G) {
    .sg_max_degree(G)
}

#' Approximate betweenness centrality
#'
#' Random (s, t) pair sampling with per-pair BFS; normalised to [0, 1].
#'
#' @param G A StreamGraph.
#' @param k Integer; number of sampled pairs (clamped to all pairs => exact).
#' @param n_threads Integer; OpenMP threads (0 = default).
#' @param seed Integer; RNG seed for reproducibility.
#' @return Numeric vector of per-node scores in [0, 1].
#' @examples
#' G <- stream_graph(); for (i in 0:4) add_edge(G, i, i + 1L)
#' betweenness_approx(G, k = 100L, seed = 42L)
#' @export
betweenness_approx <- function(G, k = 200L, n_threads = 0L, seed = 42L) {
    .sg_betweenness_approx(G, as.integer(k), as.integer(n_threads),
                           as.double(seed))
}

#' Save a graph to a binary file
#'
#' @param G A StreamGraph.
#' @param path Output file path.
#' @return Invisibly NULL.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' p <- tempfile(fileext = ".sgph")
#' save_graph(G, p)
#' @export
save_graph <- function(G, path) {
    .sg_save(G, path)
    invisible(NULL)
}

#' Load a graph from a binary file
#'
#' @param path Input file path (written by [save_graph()]).
#' @return A StreamGraph external pointer.
#' @examples
#' G <- stream_graph(); add_edge(G, 0L, 1L)
#' p <- tempfile(fileext = ".sgph"); save_graph(G, p)
#' H <- load_graph(p); n_edges(H)
#' @export
load_graph <- function(path) {
    .sg_load(path)
}

#NodeIndex: pure R, an environment with two maps

#' Create a node index
#'
#' Maps arbitrary keys to contiguous 0-indexed node IDs (insertion order).
#'
#' @return An environment of class "streamgraph_node_index".
#' @examples
#' idx <- node_index()
#' node_id(idx, "alice")
#' @export
node_index <- function() {
    idx <- new.env(parent = emptyenv())
    idx$key_to_id <- new.env(parent = emptyenv())
    idx$keys <- character(0)
    class(idx) <- "streamgraph_node_index"
    idx
}

#' Look up or assign a node ID for a key
#'
#' @param idx A node index from [node_index()].
#' @param key A key (coerced to character).
#' @return Integer ID (0-indexed); assigned on first use.
#' @examples
#' idx <- node_index()
#' node_id(idx, "alice")  # 0
#' node_id(idx, "bob")    # 1
#' @export
node_id <- function(idx, key) {
    key <- as.character(key)
    existing <- idx$key_to_id[[key]]
    if (!is.null(existing)) return(existing)
    new_id <- length(idx$keys)
    idx$key_to_id[[key]] <- new_id
    idx$keys <- c(idx$keys, key)
    new_id
}

#' Get the ID of an existing key
#'
#' @param idx A node index.
#' @param key A key (coerced to character).
#' @return Integer ID; errors if the key is absent.
#' @examples
#' idx <- node_index(); node_id(idx, "alice")
#' get_id(idx, "alice")
#' @export
get_id <- function(idx, key) {
    key <- as.character(key)
    existing <- idx$key_to_id[[key]]
    if (is.null(existing)) stop(sprintf("key not found: %s", key))
    existing
}

#' Get the key for an existing ID
#'
#' @param idx A node index.
#' @param id Integer node ID (0-indexed).
#' @return The key (character); errors if out of range.
#' @examples
#' idx <- node_index(); node_id(idx, "alice")
#' get_key(idx, 0L)
#' @export
get_key <- function(idx, id) {
    id <- as.integer(id)
    if (id < 0 || id >= length(idx$keys)) stop("id out of range")
    idx$keys[[id + 1L]]
}

#' Number of indexed keys
#'
#' @param idx A node index.
#' @return Integer.
#' @examples
#' idx <- node_index(); node_id(idx, "alice")
#' n_indexed(idx)
#' @export
n_indexed <- function(idx) {
    length(idx$keys)
}

#' Whether a key is present in the index
#'
#' @param idx A node index.
#' @param key A key (coerced to character).
#' @return Logical.
#' @examples
#' idx <- node_index(); node_id(idx, "alice")
#' has_key(idx, "alice")
#' @export
has_key <- function(idx, key) {
    !is.null(idx$key_to_id[[as.character(key)]])
}

File: src/streamgraph.cpp
#include "streamgraph.h"

#include <stdexcept>

namespace streamgraph {

StreamGraph::StreamGraph(uint32_t n_nodes_initial_, bool directed_)
    : directed(directed_), n_nodes_initial(n_nodes_initial_) {
    uint32_t cap = (n_nodes_initial_ == 0) ? STREAMGRAPH_AUTO_INITIAL_CAPACITY : n_nodes_initial_;
    adj.resize(cap);
    if (directed_) in_adj.resize(cap);
    dsu.init(cap);
    touched.assign(cap, false);

    // Auto mode starts with no referenced node; pre-allocated mode promises ids up to n_nodes_initial-1
    n_nodes_actual = n_nodes_initial_;

    // Touched nodes always have degree >= 1, so degree_histogram[0] stays 0
    degree_histogram.assign(1, 0);
}

EdgeKey StreamGraph::edge_key(uint32_t u, uint32_t v) const {
    return directed ? make_directed_key(u, v) : make_key(u, v);
}

bool StreamGraph::valid_touched_node(uint32_t u) const {
    return u < adj.size() && touched[u];
}

void StreamGraph::expand_to(uint32_t new_max) {
    if (new_max < adj.size()) return;
    size_t old_size = adj.size();
    size_t new_size = std::max(
        static_cast<size_t>(new_max) + 1,
        old_size + old_size / 2
    );
    adj.resize(new_size);
    if (directed) in_adj.resize(new_size);
    dsu.grow(static_cast<uint32_t>(new_size));
    touched.resize(new_size, false);
}

void StreamGraph::touch(uint32_t u) {
    if (!touched[u]) {
        touched[u] = true;
        ++n_touched;
        ++n_components_touched;
    }
}

void StreamGraph::histogram_move(uint32_t old_deg, uint32_t new_deg) {
    if (new_deg >= degree_histogram.size()) degree_histogram.resize(new_deg + 1, 0);
    // old_deg == 0 means the node was isolated and not yet in the histogram
    if (old_deg > 0) --degree_histogram[old_deg];
    ++degree_histogram[new_deg];
    if (new_deg > max_deg_) max_deg_ = new_deg;
}

void StreamGraph::insert_neighbour(uint32_t u, uint32_t v) {
    auto& a = adj[u];
    auto pos = std::lower_bound(a.neighbours.begin(), a.neighbours.end(), v);
    a.neighbours.insert(pos, v);
    histogram_move(a.degree, a.degree + 1);
    ++a.degree;
}

void StreamGraph::insert_in_neighbour(uint32_t u, uint32_t v) {
    auto& in = in_adj[u];
    in.insert(std::lower_bound(in.begin(), in.end(), v), v);
}

bool StreamGraph::add_edge(uint32_t u, uint32_t v) {
    if (u == v) return false;

    uint32_t hi = std::max(u, v);
    if (hi >= adj.size()) expand_to(hi);
    if (hi + 1 > n_nodes_actual) n_nodes_actual = hi + 1;

    EdgeKey key = edge_key(u, v);
    if (!edge_set.insert(key)) return false;

    touch(u);
    touch(v);

    // A directed edge whose reverse is already present adds no undirected
    // structure, so triangles must not be recounted for it.
    bool new_undirected_edge = !directed || !edge_set.contains(make_directed_key(v, u));

    insert_neighbour(u, v);
    if (directed) insert_in_neighbour(v, u);
    else insert_neighbour(v, u);

    if (dsu.unite(u, v)) --n_components_touched;

    if (new_undirected_edge) {
        common_buf.clear();
        uint32_t cnt;
        if (directed) {
            // undirected neighbourhood = out ∪ in; makes the count order-independent
            sorted_union(adj[u].neighbours, in_adj[u], union_u_buf);
            sorted_union(adj[v].neighbours, in_adj[v], union_v_buf);
            cnt = count_common_neighbours(union_u_buf, union_v_buf, common_buf);
        } else {
            cnt = count_common_neighbours(adj[u].neighbours, adj[v].neighbours, common_buf);
        }
        n_triangles += cnt;
        adj[u].triangle_count += cnt;
        adj[v].triangle_count += cnt;
        for (uint32_t w : common_buf) adj[w].triangle_count += 1;
    }

    ++n_edges;
    return true;
}

uint32_t StreamGraph::find_component(uint32_t u) {
    if (!valid_touched_node(u)) {
        throw std::out_of_range("streamgraph: node has not been touched");
    }
    return dsu.find(u);
}

bool StreamGraph::same_component(uint32_t u, uint32_t v) {
    if (!valid_touched_node(u) || !valid_touched_node(v)) return false;
    return dsu.find(u) == dsu.find(v);
}

uint32_t StreamGraph::n_components() const { return n_components_touched; }

uint32_t StreamGraph::degree(uint32_t u) const {
    if (u >= adj.size()) return 0;
    return adj[u].degree;
}

uint32_t StreamGraph::in_degree(uint32_t u) const {
    if (u >= adj.size()) return 0;
    return directed ? static_cast<uint32_t>(in_adj[u].size()) : adj[u].degree;
}

std::vector<uint32_t> StreamGraph::neighbours(uint32_t u) const {
    if (u >= adj.size()) return {};
    return adj[u].neighbours;
}

std::vector<uint32_t> StreamGraph::in_neighbours(uint32_t u) const {
    if (u >= adj.size()) return {};
    return directed ? in_adj[u] : adj[u].neighbours;
}

bool StreamGraph::has_edge(uint32_t u, uint32_t v) const {
    if (u == v) return false;
    return edge_set.contains(edge_key(u, v));
}

uint32_t StreamGraph::component_size(uint32_t u) {
    if (!valid_touched_node(u)) return 0;
    return dsu.comp_size(u);
}

uint32_t StreamGraph::n_nodes() const { return n_touched; }

double StreamGraph::density() const {
    double n = static_cast<double>(n_touched);
    if (n < 2.0) return 0.0;
    double m = static_cast<double>(n_edges);
    double possible = n * (n - 1.0);
    return directed ? (m / possible) : (2.0 * m / possible);
}

double StreamGraph::avg_degree() const {
    if (n_touched == 0) return 0.0;
    double m = static_cast<double>(n_edges);
    double n = static_cast<double>(n_touched);
    return directed ? (m / n) : (2.0 * m / n);
}

uint32_t StreamGraph::max_degree() const { return max_deg_; }

std::vector<uint32_t> StreamGraph::component_nodes(uint32_t u) {
    std::vector<uint32_t> out;
    if (u >= adj.size() || !touched[u]) return out;
    uint32_t root = dsu.find(u);
    for (uint32_t i = 0; i < n_nodes_actual; ++i)
        if (touched[i] && dsu.find(i) == root) out.push_back(i);
    return out;
}

std::vector<uint32_t> StreamGraph::component_ids() {
    std::vector<uint32_t> out;
    out.reserve(n_touched);
    for (uint32_t i = 0; i < n_nodes_actual; ++i)
        if (touched[i]) out.push_back(dsu.find(i));
    return out;
}

}
